<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Leire Abarrategui" />

<meta name="date" content="2020-10-30" />

<title>Epi-mutations detection in suspected rare disease samples</title>

<style type="text/css">
a.anchor-section {margin-left: 10px; visibility: hidden; color: inherit;}
a.anchor-section::before {content: '#';}
.hasAnchor:hover a.anchor-section {visibility: visible;}
</style>
<script>// Anchor sections v1.0 written by Atsushi Yasumoto on Oct 3rd, 2020.
document.addEventListener('DOMContentLoaded', function() {
  // Do nothing if AnchorJS is used
  if (typeof window.anchors === 'object' && anchors.hasOwnProperty('hasAnchorJSLink')) {
    return;
  }

  const h = document.querySelectorAll('h1, h2, h3, h4, h5, h6');

  // Do nothing if sections are already anchored
  if (Array.from(h).some(x => x.classList.contains('hasAnchor'))) {
    return null;
  }

  // Use section id when pandoc runs with --section-divs
  const section_id = function(x) {
    return ((x.classList.contains('section') || (x.tagName === 'SECTION'))
            ? x.id : '');
  };

  // Add anchors
  h.forEach(function(x) {
    const id = x.id || section_id(x.parentElement);
    if (id === '') {
      return null;
    }
    let anchor = document.createElement('a');
    anchor.href = '#' + id;
    anchor.classList = ['anchor-section'];
    x.classList.add('hasAnchor');
    x.appendChild(anchor);
  });
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Epi-mutations detection in suspected rare disease samples</h1>
<h4 class="author">Leire Abarrategui</h4>
<h4 class="date">2020-10-30</h4>



<div id="purpose" class="section level1">
<h1>Purpose</h1>
<p>The <code>EpiMutations</code> package provides a set of functions for epi-mutation identification in DNA methylation samples. It is designed for <code>GenomicRatioSet</code> and <code>ExpressionSet</code> data types, containing CpGs beta values.</p>
<p>The detection of outliers CpG sites is done in two steps: (1) the differentially methylated regions (DMRs) in the given sample are identified by bump-hunter approach using <code>bumphunter()</code> function. (2) CpGs in the proband sample are tested in those DMRs in order to identify regions with CpGs being outliers when comparing with the reference panel (control samples). To this end, different anomaly detection statistical approaches are used. Those include: <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/manova"><code>manova()</code></a>, <a href="https://github.com/dgarrimar/mlm"><code>mlm()</code></a>, <a href="https://rdrr.io/cran/isotree/man/isolation.forest.html"><code>isolation.forest()</code></a> and <a href="https://www.rdocumentation.org/packages/robustbase/versions/0.93-6/topics/covMcd"><code>covMcd()</code></a></p>
</div>
<div id="available-data" class="section level1">
<h1>Available data</h1>
<p>The <code>EpiMutations</code> package includes control and case datasets obtained from <a href="https://www.ncbi.nlm.nih.gov/geo/">GEO</a>.</p>
<p>The control panel corresponds to the accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE104812">GSE104812</a> which includes DNA methylation profiling (Illumina 450K BeadChip array) of 48 healthy children. The case probes were obtained from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE97362">GSE97362</a>. This dataset contains DNA methylation signatures of 8 children with CHARGE and Kabuki syndromes.</p>
<p>The available datasets correspond to CpGs beta values measured in whole blood samples. These were downloaded as <code>RGChannelSet</code> using <code>GEOquery</code> package. Then, the datasets preprocessing have been performed to remove existing Copy Number Variants (CNVs).</p>
<p>Finally, we created <code>GenomicRatioSet</code> and <code>ExpressionSet</code> type datasets, for both, control and case. These can be distinguished by the acronyms <code>grs.</code> and <code>es.</code> at the beginning of the file names.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">library</span>(EpiMutations)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="co">#Control samples</span></a>
<a class="sourceLine" id="cb1-3" title="3"><span class="co">##GenomicRatioSet</span></a>
<a class="sourceLine" id="cb1-4" title="4"><span class="kw">data</span>(grs.control.panel)</a>
<a class="sourceLine" id="cb1-5" title="5"><span class="co">##ExpressionSet</span></a>
<a class="sourceLine" id="cb1-6" title="6"><span class="kw">data</span>(es.control.panel)</a>
<a class="sourceLine" id="cb1-7" title="7"></a>
<a class="sourceLine" id="cb1-8" title="8"><span class="co">#Case samples</span></a>
<a class="sourceLine" id="cb1-9" title="9"><span class="co">##GenomicRatioSet</span></a>
<a class="sourceLine" id="cb1-10" title="10"><span class="kw">data</span>(grs.diseases)</a>
<a class="sourceLine" id="cb1-11" title="11"><span class="co">##ExpressionSet</span></a>
<a class="sourceLine" id="cb1-12" title="12"><span class="kw">data</span>(es.diseases)</a></code></pre></div>
</div>
<div id="epi-mutations-detection" class="section level1">
<h1>Epi-mutations detection</h1>
<p>The primary function of this package is <code>EpiMutations()</code>. First, it selects the DMRs. Then, it calls the function for robust outlier detection, which is selected by the user within the <code>method</code> parameter. This can be set as:</p>
<ul>
<li><code>&quot;manova&quot;</code>: Multivariate analysis of variance</li>
<li><code>&quot;mlm&quot;</code>: Multivariate linear model</li>
<li><code>&quot;iso.forest&quot;</code>: Isolation forest</li>
<li><code>&quot;Mahdist.MCD&quot;</code>: Robust Mahalanobis distance</li>
</ul>
<p>Here we illustrate how to identify outlier methylated regions in the case sample.</p>
<p>First, we create the <code>disease</code> variable with only one case sample, in order to compare it againts all the probes in the control panel:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">disease&lt;-es.diseases[,<span class="dv">4</span>]</a>
<a class="sourceLine" id="cb2-2" title="2"><span class="kw">dim</span>(disease)[<span class="dv">2</span>]</a></code></pre></div>
<pre><code>Samples 
      1 </code></pre>
<p>Then, we run <code>EpiMutations()</code> function for each method. Must be mentioned that in <code>manova</code> and <code>mlm</code> the p-value can be specified through <code>pValue.cutoff</code> parameter as a cutoff. However, The <code>iso.forest</code> uses the anomaly score indicated by <code>outlier.score</code> parameter. In the robust Mahalanobis distance, the threshold is set by chi-square distribution, but this is established by the function. Also, the minimum number of CpGs in each identified bumps can be selected through <code>num.cpgs</code>. In the below examples the default p-value, outlier score and number of CpGs are used.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co">#manova (default method)</span></a>
<a class="sourceLine" id="cb4-2" title="2"></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="kw">EpiMutations</span>(<span class="dt">diseases =</span> disease)</a></code></pre></div>
<pre><code># A tibble: 4 x 12
  chr    start    end  value  area cluster indexStart indexEnd     L clusterL
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;      &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;    &lt;int&gt;
1 X     1.19e8 1.19e8 -0.182  2.73  203453     481713   481727    15       15
2 X     1.19e8 1.19e8 -0.178  2.49  203486     481823   481836    14       15
3 X     4.67e7 4.67e7 -0.182  2.36  201663     476600   476612    13       13
4 19    5.60e7 5.60e7  0.222  2.22  102223     448922   448931    10       11
# ... with 2 more variables: sample &lt;chr&gt;, manova &lt;dbl&gt;</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="co">#Multivariate inear model</span></a>
<a class="sourceLine" id="cb6-2" title="2"></a>
<a class="sourceLine" id="cb6-3" title="3"><span class="kw">EpiMutations</span>(<span class="dt">diseases =</span> disease, <span class="dt">method =</span> <span class="st">&quot;mlm&quot;</span>)</a></code></pre></div>
<pre><code># A tibble: 2 x 12
  chr    start    end value  area cluster indexStart indexEnd     L clusterL
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;      &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;    &lt;int&gt;
1 7     2.72e7 2.72e7 0.151  2.56  175222     195992   196008    17       53
2 19    5.60e7 5.60e7 0.222  2.22  102223     448922   448931    10       11
# ... with 2 more variables: sample &lt;chr&gt;, mlm &lt;dbl&gt;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co">#Isolation forest</span></a>
<a class="sourceLine" id="cb8-2" title="2"></a>
<a class="sourceLine" id="cb8-3" title="3"><span class="kw">EpiMutations</span>(<span class="dt">diseases =</span> disease, <span class="dt">method =</span> <span class="st">&quot;iso.forest&quot;</span>)</a></code></pre></div>
<pre><code># A tibble: 17 x 12
   chr    start    end  value  area cluster indexStart indexEnd     L clusterL
   &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;      &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;    &lt;int&gt;
 1 6     3.00e7 3.00e7  0.245  3.92  163307     160113   160128    16       62
 2 6     3.00e7 3.00e7  0.239  2.63  163307     160101   160111    11       62
 3 7     2.72e7 2.72e7  0.151  2.56  175222     195992   196008    17       53
 4 X     4.64e7 4.64e7 -0.169  2.53  201651     476557   476571    15       15
 5 X     1.54e8 1.54e8 -0.179  2.50  204615     484935   484948    14       21
 6 X     1.19e8 1.19e8 -0.178  2.49  203486     481823   481836    14       15
 7 X     7.74e7 7.74e7 -0.188  2.25  202731     479666   479677    12       15
 8 19    5.60e7 5.60e7  0.222  2.22  102223     448922   448931    10       11
 9 X     4.75e7 4.75e7 -0.173  2.08  201748     476874   476885    12       17
10 5     1.26e8 1.26e8 -0.187  2.06  155596     140640   140650    11       14
11 X     4.89e7 4.89e7 -0.169  2.02  201899     477344   477355    12       12
12 X     4.88e7 4.88e7 -0.197  1.97  201875     477265   477274    10       10
13 X     1.54e8 1.54e8 -0.179  1.79  204603     484886   484895    10       11
14 X     1.06e8 1.06e8 -0.173  1.73  203136     480858   480867    10       11
15 X     1.95e7 1.95e7 -0.169  1.69  201076     475042   475051    10       11
16 X     1.50e8 1.50e8 -0.166  1.66  204170     483761   483770    10       12
17 X     4.84e7 4.84e7 -0.162  1.62  201808     477015   477024    10       17
# ... with 2 more variables: sample &lt;chr&gt;, iso &lt;dbl&gt;</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co">#Robust mahalanobis distance</span></a>
<a class="sourceLine" id="cb10-2" title="2"></a>
<a class="sourceLine" id="cb10-3" title="3"><span class="kw">EpiMutations</span>(<span class="dt">diseases =</span> disease, <span class="dt">method =</span> <span class="st">&quot;Mahdist.MCD&quot;</span>)</a></code></pre></div>
<pre><code># A tibble: 4 x 12
  chr    start    end  value  area cluster indexStart indexEnd     L clusterL
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;      &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;    &lt;int&gt;
1 6     3.00e7 3.00e7  0.239  2.63  163307     160101   160111    11       62
2 7     2.72e7 2.72e7  0.151  2.56  175222     195992   196008    17       53
3 19    5.60e7 5.60e7  0.222  2.22  102223     448922   448931    10       11
4 5     1.26e8 1.26e8 -0.187  2.06  155596     140640   140650    11       14
# ... with 2 more variables: sample &lt;chr&gt;, MahMCD &lt;lgl&gt;</code></pre>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
