---
title: "Detection of epimutations with state of the art methods in methylation data"
author:
- name: Leire Abarrategui Martinez
  affiliation:
  - &isglobal Bioinformatics Research Group in Epidemiolgy (BRGE), Barcelona Insitute for Global Health (ISGlobal)
  - &Faculty of Medical Sciences, Newcastle University, Newcastle-Upon-Tyne, UK
  - &Autonomous University of Barcelona (UAB)
  email: leire.abarrategui-martinez@newcastle.ac.uk
- name: Carles Hernandez-Ferrer
  affiliation:
  - &Centro Nacional de Análisis Genómico (CNAG-CRG), Center for Genomic, Regulation
  - &Barcelona Institute of Science and Technology (BIST), Barcelona, Catalonia, Spain
  email: carles.hernandez@cnag.crg.eu
- name: Carlos Ruiz-Arenas
  affiliation:
  - &Centro de Investigación Biomédica en Red de Enfermedades Raras (CIBERER), Barcelona, Spain
  - &Universitat Pompeu Fabra (UPF), Barcelona, Spain
  email: carlos.ruiza@upf.edu
- name: Juan R. Gonzalez
  affiliation:
  - &isglobal Bioinformatics Research Group in Epidemiolgy (BRGE), Barcelona Insitute for Global Health (ISGlobal)
  - &uab Department of Mathematics, Autonomous University of Barcelona (UAB)
  email: juanr.gonzalez@isglobal.org
date: "`r Sys.Date()`"
package: "`r pkg_ver('epimutacions')`"
abstract: >
    An introductory guide to analysing DNA methylation data with R package `epimutacions`. The areas covered in this document are: installing the     package, loading the data, identifying the epimutations and results visualisation.  

output:
  BiocStyle::html_document
bibliography: refs.bib
vignette: >
  %\VignetteIndexEntry{Detection of epimutations with state of the art methods in methylation data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(comment="", warning = FALSE, message = FALSE, cache = TRUE)
```




# Introduction

Rare diseases are pathologies with a low prevalence (< 1 per 2,000 people). Most of these pathologies have an onset during childhood and a strong genetic etiology. Consequently, rare disease diagnosis has relied on identifying genetic and genomic mutations that can cause the disease. Although these variants have provided a diagnosis for many patients and families, around 60% of the cases remained undiagnosed [@lionel2018improved]. Aberrant methylation can be an underlying cause of undiagnosed patients, either as a primary event (a.k.a. epimutation) or as a functional consequence of chromatin dysregulation by genetic or environmental agents (a.k.a. episignature). Epimutations are the cause of some rare diseases, such as Prader-Willi, Angelman or Beckwith-Wiedemann syndromes [@aref2019] and some human malformations [@serra2015dna]. Syndrome-specific episignatures are increasingly defined as biomarkers for a growing number of disorders [@aref2019]. Therefore, tools to detect epimutations and episignatures should be made available to the rare disease community and included in standardized analysis workflows.


The package `epimutacions ` provides tools for analysing epivariants using multiple outlier detection  approaches. In addition, functions to plot and annotate 
the identified epimutations. The full `epimutacions ` user´s guide is available in this vignette. 

The name of the package is `epimutacions` (pronounced `ɛ pi mu ta 'sj ons`) which means epimutations in Catalan, a language from the northeast of Spain.


# Setup

## Installing the package

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("epimutacions")
```


## Loading libraries

```{r, message=FALSE}
library(epimutacions)
```

## Preparing dataset

We defined 40408 candidate epimutation regions (found in `candRegsGR`) based on the 450K  array design. As CpGs are not equally distributed along the genome, only CpGs closer to other CpGs can form an epimutation. We cluster the CpGs that were closer than 1000Kb to their closest CpG.

Next, we mapped epimutations regions to hg38. We selected those regions whose mapping between hg19 and hg38 builds was 100%. Also, we overlapped epimutation regions with ENCODE cREs v13. 


```{r}
data("candRegsGR")
candRegsGR
```

In addition, the package includes a small `GenomicRatioSet` class dataset (`methy` `r data("methy", package = "epimutacions")`) containing the DNA methylation profiles from a total of  `r ncol(methy)` individuals, `r ncol(methy[,methy$status == "case"])` cases and `r ncol(methy[,methy$status == "control"])` controls. The DNA methylation profiles were generated using the Illumina 450k Human Methylation BeadChip. The data were obtained from [Gene Expression Omnibus (GEO)](https://www.ncbi.nlm.nih.gov/geo/) and adapted for the package usage. 


```{r}
data("methy")
methy

table(methy$status)
```

# Preprocessing
The preprocessing in `epimutacions` package is done by `epi_preprocess()` function. It contains 6 preprocessing method corresponding to minfi package:

+---------------+-----------------------------+----------------------------------------------------------------------+
| Method        |  Function                   | Description                                                          |
+===============+=============================+======================================================================+
| `raw`         | `preprocessRaw`             | Converts the Red/Green channel for an Illumina methylation           |  
|               |                             |  array into methylation signal, without using any normalization      |
+---------------+-----------------------------+----------------------------------------------------------------------+
| `illumina  `  | `preprocessIllumina`        |  Implements preprocessing for Illumina methylation                   |
|               |                             |      microarrays as used in Genome Studio                            |
+---------------+-----------------------------+----------------------------------------------------------------------+
|`swan`         | `preprocessSWAN `           |Subset-quantile Within Array Normalisation (SWAN). It allows Infinium |
|               |                             | I and II type probes on a single array to be normalized together     |
+---------------+-----------------------------+----------------------------------------------------------------------+
|  `quantile`   |`preprocessQuantile`         |Implements stratified quantile normalization preprocessing for        |                
|               |                             |Illumina methylation microarrays                                      |
+---------------+-----------------------------+----------------------------------------------------------------------+
| `noob`        | `preprocessNoob`            | Noob (normal-exponential out-of-band) is a background correction     | 
|               |                             | method with dye-bias normalization for                               |
|               |                             | Illumina Infinium methylation arrays                                 |
+---------------+-----------------------------+----------------------------------------------------------------------+
| `funnorm`     | `preprocessFunnorm `        | Functional normalization (FunNorm) is a between-array                |
|               |                             | normalization method for the Illumina Infinium                       |
|               |                             | HumanMethylation450 platform                                         | 
+---------------+-----------------------------+----------------------------------------------------------------------+

In addition, `norm_parameters()` function contains  the parameters for each method described above: 

+----------+-----------------------+-----------------------------------------------------------------------------+
| Method   |Parameters             |Description                                                                  |
+==========+=======================+=============================================================================+
|`illumina`|`bg.correct`<br>       |Performs background correction<br>                                           |
|          |`normalize`<br>        |Performs controls normalization<br>                                          |        
|          |`reference`            |The reference array for control normalization                                |
+----------+-----------------------+-----------------------------------------------------------------------------+
|`quantile`|`fixOutliers`<br>      | Low outlier Meth and Unmeth signals will be fixed<br>                       |                       
|          |`removeBadSamples`<br> | Remove bad samples<br>                                                      |
|          |`badSampleCutoff`<br>  | The cutoff to label samples as 'bad'<br>                                    |
|          |`quantileNormalize`<br>| Performs quantile normalization<br>                                         |
|          | `stratified`          | performs quantile normalization within region strata<br>                    |
|          |`mergeManifest`<br>    | Merged to the output the information in the associated manifest package<br> |
|          |`sex`                  | Sex of the samples                                                          |
+----------+-----------------------+-----------------------------------------------------------------------------+
|`noob`    | `offset`<br>          | Offset for the normexp background correct<br>                               |
|          | `dyeCorr`<br>         | Performs dye normalization<br>                                              |  
|          | `dyeMethod`           | Dye bias correction to be done                                              |
+---------------+------------------+-----------------------------------------------------------------------------+
|`funnorm` | `nPCs `<br>           | The number of principal components from the control probes<br>              |       
|          | `sex`<br>             | Sex of the samples<br>                                                      |      
|          | `bgCorr`<br>          | Performs NOOB background correction prior to functional normalization<br>   |
|          | `dyeCorr`<br>         | Performs dye normalization<br>                                              |
|          | `keepCN`              | Keeps copy number estimates                                                 |
+---------------+-----------------------------+------------------------------------------------------------------+

Invoking `norm_parameters()` with no arguments returns a list of the default settings for each method: 

```{r}
norm_parameters()
```

If we do not want to perform the background correction in `illumina` method we can do as following: 

```{r}
parameters <- norm_parameters(illumina = list("bg.correct" = FALSE))
parameters$illumina$bg.correct
```
# Epimutations

## Methods included for epimutations identification

The `epimutacions` package includes 6 methods to detect epimutations. Two of these methods were described elsewhere before [@barbosa2018; @aref2019], and the other three are adaptations of one of them:

| Method | Short description |
|:------:|:------------------|
| `barbosa`     | Previously described [@barbosa2018]                                                                     |
| `manova`      | Previously described [@aref2019]                                                                        |
| `mlm`         | Adapted from [@aref2019] using multivariate linear methods to identify outliers                         |
| `iso.forest`  | Adapted from [@aref2019] using isolation forest to detect outliers                                      |
| `mahdist.mcd` | Adapted from [@aref2019] using robust Mahalanobis distance with a custom threshold to identify outliers |
| `beta`        |Identifies epimutations based on a beta distribution                                                     |

### Function parameters for each method

The `epi_parameters()` function  is  useful to set the value of the parameters for the above methods. 
In the following table each argument is described: 

+---------------+-----------------------------+----------------------------------------------------------------------+
| Method        | Parameter                   | Description                                                          |
+===============+=============================+======================================================================+
| `manova`<br>  | `pvalue_cutoff`             | The threshold p-value to select which CpG regions are outliers       |
|  `mlm`<br>    |                             |                                                                      |
|  `beta`       |                             |                                                                      |
+---------------+-----------------------------+----------------------------------------------------------------------+
| `iso.forest`  | `outlier_score_cutoff`<br>  | The threshold to select which CpG regions are outliers<br>           |
|               | `ntrees`                    | The number of binary trees to build for the model                    |
+---------------+-----------------------------+----------------------------------------------------------------------+
|`mahdist.mcd`  | `nsamp`                     | The number of subsets used for initial estimates in the MCD          |
+---------------+-----------------------------+----------------------------------------------------------------------+
|  `barbosa`    |`window_sz`<br>              |The maximum distance between CpGs to be considered in the same DMR<br>|                
|               |`offset_mean`/`offset_abs`   | The upper and lower threshold to consider a CpG an outlier           |
+---------------+-----------------------------+----------------------------------------------------------------------+
| `beta`        | `pvalue_cutoff`<br>         | The minimum p-value to consider a CpG an outlier<br>                 |
|               | `diff_threshold`            | The minimum methylation difference between the CpG and the           |
|               |                             |  mean methylation to consider a position an outlier                  |
+---------------+-----------------------------+----------------------------------------------------------------------+

Invoking `epi_parameters()` with no arguments returns a list of the default settings for each method: 

```{r}
epi_parameters()
```

If we want to change the p-value cut off for `manova` we can do as following: 

```{r}
parameters <- epi_parameters(manova = list("pvalue_cutoff" = 0.01))
parameters$manova$pvalue_cutoff
```

### Methods from literature

We called `"barbosa"` to the method described in [@barbosa2018].  Briefly, this algorithm check for each CpG, if the proband's measurement is an outlier. Then, it calls an epimutation to those regions where 3 contiguous CpGs are outliers and they are separated by less than 500 base pairs.

```{r, epi_brb, message=FALSE}
epi_brb <- epimutations(methy, method = "barbosa")
```

We called `"manova"` to the implementation of [@aref2019]. This algorithm starts by running `bumphunter`, a method to detect DMRs (Differentially Methylated Regions) [@peters2015novo], comparing the proband versus the reference group. This first step results in a list of DMRs, regions where the proband has different methylation than the reference group. Then, each DMR is tested for outlier using a MANOVA (Multivariate ANalysis Of VAriance). DMRs are then subset based on the F-statistic magnitude and the mean difference between proband and reference group DNA methylation.

```{r, epi_mvo, message=FALSE, warning=FALSE}
epi_mvo <- epimutations(methy, method = "manova")
```

### Altered methods from literature

From [@aref2019] implementation, we proposed three different approaches. These three approaches run the `bumphunter` step but test DMR for outliers using different approaches: (1) `"mlm"`, which uses a multivariate linear model; (2) `"isoforest"`, that uses isolation forest; and (3) `"mahdistmcd"`, which uses robust Mahalanobis distance.

```{r, epi_mlm, message=FALSE, warning=FALSE}
epi_ml <- epimutations(methy, method = "mlm")
```

```{r, epi_iso, message=FALSE, warning=FALSE}
epi_iso <- epimutations(methy, method = "isoforest")
```

```{r, epi_mcd, message=FALSE, warning=FALSE}
epi_mcd <- epimutations(methy, method = "mahdistmcd")
```

### Other methods

We implemented `"beta"` method which  models the DNA methylation data using a beta distribution.

```{r, epi_beta, message=FALSE, warning=FALSE}
epi_beta <- epimutations(methy, method = "beta")
```

## Results description

The two methods from literature returned `r nrow(epi_brb)` and `r nrow(epi_mvo)` epimutations as we can see:

```{r, show_brb_mvo_1}
nrow(epi_brb) # Results from Barbosa et. al. 2018
nrow(epi_mvo) # Results from Aref-Eshghi et. al. 2019
```

All the implemented methods return the same `tibble` although the meaning of some of the columns is slightly different. Let's see one of the outputs from the two methods from literature:

```{r, show_brb_mvo_2}
rbind(epi_brb[1, ], epi_mvo[1, ])
```

+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| Column name           | Description                                                                                                           |
+=======================+=======================================================================================================================+
| `epi_id`              | systematic name for each epimutation identified                                                                       |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `sample`              | The name of the sample containing that epimutation                                                                    |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `chromosome`          | The location of the epimutation                                                                                       |
|`start`                |                                                                                                                       |
|`end`                  |                                                                                                                       |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `sz`                  | The window's size of the event                                                                                        |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `cpg_n`               | The number of CpGs in the epimutation                                                                                 |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `cpg_n`               | The names of CpGs in the epimutation                                                                                  |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `outlier_score`       | For method `manova` it provides the approximation to F-test and the Pillai score, separated by `/`<br>                |
|                       | For method `mlm` it provides the approximation to F-test and the R2 of the model, separated by `/`<br>                |
|                       | For method `isoforest` it provides the magnitude of the outlier score.<br>                                            |
|                       | For method `beta` it provides the mean p-value of all GpGs in that DMR<br>                                            |
|                       | For methods `barbosa` and `mahdistmcd` it is filled with `NA`.                                                        |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `pvalue`              | For methods `manova` and  `mlm` it provides the p-value obtained from the model.<br>                                  |
|                       | For method `barbosa`, `isoforest`, `beta` and `mahdistmcd` it is filled with `NA`.                                    |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+ 
| `outlier_direction`   | Indicates the direction of the outlier with "hypomethylation" and "hypermethylation".<br>                             | 
|                       | For `manova`, `mlm`, `isoforest`, and `mahdistmcd` it is computed from the values obtained from `bumphunter`.<br>     |
|                       | For `beta` is computed from the p value for each CpG using `diff_threshold` and `pvalue_threshold` arguments.<br>     |
|                       | For `barbosa` it is computed from the location of the sample in the reference distribution (left vs. right outlier).  |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `adj_pvalue`          | For methods `manova` and  `mlm` it provides the adjusted p-value with                                                 |
|                       | Benjamini-Hochberg based on the total number of regions detected by Bumphunter.<br>                                   |
|                       | For method `barbosa`, `isoforest`, `mahdistmcd` and `beta` it is filled with `NA`.                                    |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+ 
| `epi_region_id`       | Name of the epimutation region as defined in `candRegsGR`.                                                            |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `CRE`                 | cREs (cis-Regulatory Elements) as defined by ENCODE overlapping the epimutation region.                               |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+
| `CRE_type`            | Type of cREs (cis-Regulatory Elements) as defined by ENCODE.                                                          |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------+

## Epimutations annotations

The `epimutacions` package also includes the `annotate_epimutations` function dedicated to enrich the epimutations identified by the previously described methods:

```{r, ann, message=FALSE, tidy = TRUE}
rst_mvo <- annotate_epimutations(epi_mvo)
```

```{r ann results, eval = FALSE}
rst_mvo[1:2, c(1, 12:14)]
```

```{r echo = FALSE}
library(knitr)
kable(rst_mvo[1:2, c(1, 12:14)], caption = "epimutations annotation")
```


## Epimutation visualization

The  visualization approach locates the epimutations along the genome. The function `plot_epimutations` plots the methylation values of the  individual with the epimutation in red, the control samples in dashed black lines and population mean in blue:  
```{r}
plot_epimutations(as.data.frame(epi_mvo[1,]), methy)
```

Furthermore, it includes the gene annotations in the regions in which the epivariation is located.  This can be achieved by using the argument  `gene_annot == TRUE`: 

```{r, plot_mvo_genes_annot}
plot_epimutations(as.data.frame(epi_mvo[1,]), methy, genes_annot = TRUE)
```


Also, it is possible to plot the chromatin marks H3K4me3, H3K27me3  and H3K27ac by setting the argument `regulation = TRUE`:

* **H3K4me3**: commonly associated with the activation of transcription of nearby genes.
* **H3K27me3**: is used in epigenetics to look for inactive genes.
* **H3K27ac**: is associated with the higher activation of transcription and therefore defined as an active enhancer mark


```{r, plot_mvo_regulation}
plot_epimutations(as.data.frame(epi_mvo[1,]), methy, regulation = TRUE)
```


# Acknowledgements

We acknowledge the organizers of the [European BioHackathon 2020](https://www.biohackathon-europe.org/) for their support.

All the team members of *Project #5* for the contribution to this package: 

| Name | Surname | ORCID | Affiliation | Team |
|:-----|:--------|:-----:|:------------|:-----|
| Leire | Abarrategui | 0000-0002-1175-038X | Faculty of Medical Sciences, Newcastle University, Newcastle-Upon-Tyne, UK; Autonomous University of Barcelona (UAB), Barcelona, Spain | Development |
| Lordstrong | Akano | 0000-0002-1404-0295 | College of Medicine, University of Ibadan | Development |
| James | Baye | 0000-0002-0078-3688 | Wellcome/MRC Cambridge Stem Cell Institute, University of Cambridge, Cambridge CB2 0AW, UK; Department of Physics, University of Cambridge, Cambridge CB2 3DY, UK | Development |
| Alejandro | Caceres | - | ISGlobal, Barcelona Institute for Global Health, Dr Aiguader 88, 08003 Barcelona, Spain; Centro de Investigación Biomédica en Red en Epidemiología y Salud Pública (CIBERESP), Madrid, Spain | Development |
| Carles | Hernandez-Ferrer | 0000-0002-8029-7160 | Centro Nacional de Análisis Genómico (CNAG-CRG), Center for Genomic, Regulation; Barcelona Institute of Science and Technology (BIST), Barcelona, Catalonia, Spain | Development |		
| Pavlo | Hrab | 0000-0002-0742-8478 | Department of Genetics and Biotechnology, Biology faculty, Ivan Franko National University of Lviv | Validation |
| Raquel | Manzano | 0000-0002-5124-8992 | Cancer Research UK Cambridge Institute; University of Cambridge, Cambridge, United Kingdom | Reporting |
| Margherita | Mutarelli | 0000-0002-2168-5059 | Institute of Applied Sciences and Intelligent Systems (ISASI-CNR) | Validation |
| Carlos | Ruiz-Arenas | 0000-0002-6014-3498 | Centro de Investigación Biomédica en Red de Enfermedades Raras (CIBERER), Barcelona, Spain; Universitat Pompeu Fabra (UPF), Barcelona, Spain | Reporting |

# References
